### PR TITLE
Specify layer and group IDs in a less magical way when setting order in the Layers Panel

### PR DESCRIPTION
**Pull request by MattF-NSIDC**
## Description

title

TODO:

- [x] unit test
- [x] lockfile

## Checklist

If an item on this list is done _or not needed_, simply check it with `[x]`.

- [x] Config lockfile updated (`inv config.export > qgreenland/config/cfg-lock.json`)
- [x] Version bumped if needed (`bumpversion (major|minor|patch|prerelease|build`)
- [x] CHANGELOG.md updated
- [x] Documentation updated if needed
- [x] New unit tests if needed

### CODE REVIEW

**Comment by MattF-NSIDC** [comment]
I think ReadTheDocs is failing because it's using an older version of Python that doesn't have support for the Union operator `|`. Not sure yet on GitHub Actions...

File: doc/discussion/pydantic-and-mypy.md
Comment (trey-stafford) [inline doc/discussion/pydantic-and-mypy.md:1]: Nice explanation!

File: qgreenland/models/config/layer_group.py [hunk 1]
```diff
@@ -1,19 +1,104 @@
...
+    Based on docs: https://docs.pydantic.dev/1.10/usage/types/#classes-with-__get_validators__
+
+    TODO: Update other models to replace string layer identifiers with this class?
+    """
+
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    # TODO: __modify_schema__?
```
Comment (trey-stafford) [inline qgreenland/models/config/layer_group.py:23]: What does this do? More context or remove maybe?

File: qgreenland/models/config/layer_group.py [hunk 2]
```diff
@@ -1,19 +1,104 @@
...
+    This is not expected to be globally unique, only locally within their parent group.
+    For example, there may be two group _paths_:
+
+        * Foo/January
+        * Bar/January
+
+    `Foo`, `Bar`, `January` are all LayerGroupIdentifiers, but `January` is not globally
+    unique, only locally (i.e. there can't be two `January`s in `Foo`).
+
+    TODO: Update other models to replace string layer group identifiers with lists of
```
Comment (trey-stafford) [inline qgreenland/models/config/layer_group.py:55]: I was thinking the same thing!

File: qgreenland/models/config/layer_group.py [hunk 3]
```diff
@@ -1,19 +1,104 @@
...
+    this class?
+    TODO: Dedup methods
```
Comment (trey-stafford) [inline qgreenland/models/config/layer_group.py:57]: Remove? Which methods need de-duped?

File: qgreenland/util/config/compile.py [hunk 4]
```diff
@@ -73,4 +73,7 @@ def compile_cfg(
             layer_tree=compiled_layer_tree,
         )
     except Exception as e:
-        raise exc.QgrConfigCompileError(f"Failed to compile config. {e}") from e
+        raise exc.QgrConfigCompileError(
+            f"Failed to compile config at '{config_dir}'. NOTE: Follow the traceback up"
+            f" to find the file at fault.\n{e}"
```
Comment (trey-stafford) [inline qgreenland/util/config/compile.py:78]: :rocket:

File: qgreenland/util/tree.py [hunk 5]
```diff
@@ -206,12 +208,14 @@ def _manual_ordering_strategy(
...
-            else:
-                matcher = lambda x, s=s: isinstance(x, Path) and x.name == s
+            if isinstance(s, LayerIdentifier):
+                matcher = lambda x, s=s: isinstance(x, Layer) and x.id == str(s)
+                thing_desc = f'layer id "{s}"'
+            elif isinstance(s, LayerGroupIdentifier):
+                matcher = lambda x, s=s: isinstance(x, Path) and x.name == str(s)
                 thing_desc = f'group/directory "{s}"'
+            else:
+                raise TypeError("This should never happen")
```
Comment (trey-stafford) [inline qgreenland/util/tree.py:218]: :trollface:

File: qgreenland/test/util/config/test_compile_validation.py [hunk 6]
```diff
@@ -4,17 +4,26 @@
...
+    [
+        (
+            TEST_CONFIG_W_MISSING_SETTINGS_DIR,
+            "Found the following items on disk but not in ordered set",
+        ),
+        (TEST_CONFIG_W_EXTRA_SETTINGS_DIR, "Expected to find layer id"),
+        (TEST_CONFIG_W_DUPES_DIR, "Duplicate layer_ids found"),
+        (TEST_CONFIG_W_STRING_ORDER_VALUES, "Must be explicitly initialized"),
+    ],
+)
```
Comment (MattF-NSIDC) [inline qgreenland/test/util/config/test_compile_validation.py:26]: These tests now check for a specific substring in the error message.

File: qgreenland/util/config/compile.py (hunk 4 above)
Comment (MattF-NSIDC) [inline qgreenland/util/config/compile.py:78]: I struggled for too long with the Pydantic error messages, which don't point to a file or line number. If you look really far up the traceback, you can find it.

**Review by trey-stafford — APPROVED** [review]
Nice, this is definitely better! A couple of comments pointing out TODOs but otherwise looks good to me.

File: qgreenland/models/config/layer_group.py (hunk 1 above)
Comment (MattF-NSIDC) [reply qgreenland/models/config/layer_group.py]: e27f2941

File: qgreenland/models/config/layer_group.py (hunk 3 above)
Comment (MattF-NSIDC) [reply qgreenland/models/config/layer_group.py]: 07e07f96

**Pull request merged**